### DISCUSSION TITLE
Using custom schemas with the JavaScript client

### DISCUSSION THREAD

**Discussion by SunshineyDay** [supabase/supabase#21511] [Questions] [answered]
I come from a SQL Server background where schemas were frequently used as an organisational tool to segregate areas of your database, like namespacing. 

I'm working on a Supabase project at the moment that I initially felt would benefit from similar usage of schemas. I have a load of tables in the public schema, some of which will be populated with data by an import process in my application. 

To aid the import process I've created a number of tables under a _staging_ schema. The idea being that I will initially import to those tables, put some UI over the data so that the user can confirm the import, and then move the data from the staging tables to their public equivalent. 

Having read the Supabase docs about [custom schemas](https://supabase.com/docs/guides/api/using-custom-schemas) I'm now reconsidering this approach as it looks like I can only set the schema on the entire JavaScript client, not on the individual query/mutation.

Perhaps I'm looking at this with my SQL Server glasses on, but I'm just wondering if it would be too much of a stretch to be able to specify the fully qualified schemaName.tableName on a per query basis in the Supabase client's .from() clause so that we can freely move around schemas in our queries and mutations? Is this not a very Postgresy thing to do?

**Comment by GaryAustin1** [comment, marked as answer]
You can set the schema on each database request with .schema('schema').
https://supabase.com/docs/reference/javascript/select?example=switching-schemas-per-query

Edit: note this only allows 1 schema to be used for the entire request.  PostgREST API does not have ability to handle multiple schemas in a single request.

**Reply by SunshineyDay** [reply]
Thanks for this - I didn't see this option (is it a recent addition?).

This does address my immediate use case. It would still be nice to be able to specify the schema for each table as part of the .from() clause, but without knowing anything about how PostgREST works under the covers this may well be easier said than done.

For the Supabase docs maintainers it would be worth mentioning .schema() in the [using custom schemas page](https://supabase.com/docs/guides/api/using-custom-schemas).

Thanks again for the quick reply.

**Reply by GaryAustin1** [reply]
That is definitely limited by PostgREST so looking to their repository is the thing to do if you want that as a feature or see the likely requests for it already.  Right now it takes a header to set the schema for the operation.

**Reply by Donnerstagnacht** [reply]
Let's say I have the schema public, myschema1 and myschema2. 

What would be the recommended syntax use myschema1 and myschema2 in two different api calls?

Do I have to mention all schemas in thr create client statement? If yes, how?

**Reply by GaryAustin1** [reply]
The REST API only works on one schema at a time for all tables in the query.  
So .schema('x') is typically the way to go.  
You can set the schema in createClient also, but it is still just one schema that will take the place of public.

**Reply by Donnerstagnacht** [reply]
So, basically, I can use the create client to define a default schema which I can use without the schema() clause and I can use all other exposed schemas, too, but in these cases I always have to use the schema() clause in the api call?

Did I understand it correct?

**Reply by GaryAustin1** [reply]
Sounds right.  But the only schema used for all tables in the REST call is either public, or one in createClient, or the one used in .schema(), in that order of override.
